### ISSUE TITLE
Strange memory allocations when running --batchSystem gridengine on Openstack VMs

### ISSUE THREAD

**Issue by jsaydo** [DataBiosphere/toil#1713]
Hello there,

I'm new here and having some trouble with getting my simple (two times scattered) CWL-workflow to run on the gridEngine batch system. I actually don't really understand what's going wrong, but what struck me was the debug messages about memory allocations which are also reported in #1704

```
Job: Y/E/joboaJBkN has 10 successors to schedule
Issuing the command: /vol/spool/venv/bin/_toil_worker file:/vol/spool/toil-jobstore e/m/jobN1BLAK with memory: 4294967296, cores: 1, disk: 2147483648
Acquiring 4294967296 bytes of memory from a pool of 8372887552.
Acquiring 10 fractional cores from a pool of 40 to satisfy a request of 1.000000 cores
Issued job 'file:///vol/spool/cwl-kraken/workflow/kraken.cwl.yml' kraken e/m/jobN1BLAK with job batch system ID: 2 and cores: 1, disk: 2.0 G, and memory: 4.0 G
Issuing the command: /vol/spool/venv/bin/_toil_worker file:/vol/spool/toil-jobstore c/u/jobg67ahx with memory: 4294967296, cores: 1, disk: 2147483648
Acquiring 4294967296 bytes of memory from a pool of 4077920256.
Issued job 'file:///vol/spool/cwl-kraken/workflow/kraken.cwl.yml' kraken c/u/jobg67ahx with job batch system ID: 3 and cores: 1, disk: 2.0 G, and memory: 4.0 G
Issuing the command: /vol/spool/venv/bin/_toil_worker file:/vol/spool/toil-jobstore D/c/jobA9lKVS with memory: 4294967296, cores: 1, disk: 2147483648
Acquiring 4294967296 bytes of memory from a pool of 4077920256.
Issued job 'file:///vol/spool/cwl-kraken/workflow/kraken.cwl.yml' kraken D/c/jobA9lKVS with job batch system ID: 4 and cores: 1, disk: 2.0 G, and memory: 4.0 G
Issuing the command: /vol/spool/venv/bin/_toil_worker file:/vol/spool/toil-jobstore R/s/jobDgix0U with memory: 4294967296, cores: 1, disk: 2147483648
Acquiring 4294967296 bytes of memory from a pool of 4077920256.
Issued job 'file:///vol/spool/cwl-kraken/workflow/kraken.cwl.yml' kraken R/s/jobDgix0U with job batch system ID: 5 and cores: 1, disk: 2.0 G, and memory: 4.0 G
Issuing the command: /vol/spool/venv/bin/_toil_worker file:/vol/spool/toil-jobstore Q/M/jobOMQ_F3 with memory: 4294967296, cores: 1, disk: 2147483648
Acquiring 4294967296 bytes of memory from a pool of 4077920256.
Issued job 'file:///vol/spool/cwl-kraken/workflow/kraken.cwl.yml' kraken Q/M/jobOMQ_F3 with job batch system ID: 6 and cores: 1, disk: 2.0 G, and memory: 4.0 G
Issuing the command: /vol/spool/venv/bin/_toil_worker file:/vol/spool/toil-jobstore p/N/jobFmoDnl with memory: 4294967296, cores: 1, disk: 2147483648
Acquiring 4294967296 bytes of memory from a pool of 4077920256.
Issued job 'file:///vol/spool/cwl-kraken/workflow/kraken.cwl.yml' kraken p/N/jobFmoDnl with job batch system ID: 7 and cores: 1, disk: 2.0 G, and memory: 4.0 G
Issuing the command: /vol/spool/venv/bin/_toil_worker file:/vol/spool/toil-jobstore x/W/jobbpAjla with memory: 4294967296, cores: 1, disk: 2147483648
Issued job 'file:///vol/spool/cwl-kraken/workflow/kraken.cwl.yml' kraken x/W/jobbpAjla with job batch system ID: 8 and cores: 1, disk: 2.0 G, and memory: 4.0 G
Acquiring 4294967296 bytes of memory from a pool of 4077920256.

[...]

Finished job. self.coreFractions ~ 40 and self.memory ~ 8372887552
Ran jobID: 4 with exit value: 0
Job ended successfully: 'file:///vol/spool/cwl-kraken/workflow/kraken.cwl.yml' kraken D/c/jobA9lKVS
Cleaning the predecessors of D/c/jobA9lKVS
Acquiring 10 fractional cores from a pool of 40 to satisfy a request of 1.000000 cores
Could not acquire enough (memory) to run job. Requested: (4294967296), Avaliable: 4077920256. Sleeping for 10s.
Could not acquire enough (memory) to run job. Requested: (4294967296), Avaliable: 4077920256. Sleeping for 10s.
Could not acquire enough (memory) to run job. Requested: (4294967296), Avaliable: 4077920256. Sleeping for 10s.
Could not acquire enough (memory) to run job. Requested: (4294967296), Avaliable: 4077920256. Sleeping for 10s.
Could not acquire enough (memory) to run job. Requested: (4294967296), Avaliable: 4077920256. Sleeping for 10s.
Could not acquire enough (memory) to run job. Requested: (4294967296), Avaliable: 4077920256. Sleeping for 10s.
Could not acquire enough (memory) to run job. Requested: (4294967296), Avaliable: 4077920256. Sleeping for 10s.
Got message from job at time 06-13-2017 15:28:48: CACHE: Creating a harbinger file for (d/D/tmp3mPbDI.tmp). 
Got message from job at time 06-13-2017 15:28:48: CACHE: Deleting the harbinger file for (d/D/tmp3mPbDI.tmp)
Got message from job at time 06-13-2017 15:28:48: Job 'file:///vol/spool/cwl-kraken/workflow/kraken.cwl.yml' kraken D/c/jobA9lKVS used 0.50% (10.3 MB [10780672B] used, 2.0 GB [2147483648B] requested) at the end of its run.
Got message from job at time 06-13-2017 15:28:48: Successfully deleted local copies of file with ID 'd/D/tmp3mPbDI.tmp'.
Got message from job at time 06-13-2017 15:28:48: Successfully deleted local copies of file with ID '2/T/tmp2SWGQc.tmp'.
Received Toil worker log. Disable debug level logging to hide this output
```

My workflow consists of 2 steps which are run 10 times each. So after the first steps 4294967296 byte  RAM are allocated from the 8 GB RAM in total, so there are just 4077920256 byte left for the next step. But why is the RAM not freed? Or do I read the message wrong?

Thanks in advance!

## Further information:

I've set the gridengine specific variables:
```
export TOIL_GRIDENGINE_PE='multislot'
export TOIL_GRIDENGINE_ARGS='-q main.q'
```
and set the jobstore to be on a shared file system at `/vol/spool/toil-jobstore` and the workdir to be on a non-shared file system at `/tmp`.

I'm telling CWL to use 1 core and 4 GB of RAM for each tool being used (defined in each CWL CommandLineTool):

parallel workflow:
```
cwlVersion: v1.0
class: Workflow

inputs:
  fastaFile: File[]
  krakenOutputFileName: string[]
  krakenTranslateOutputFileName: string[]
  krakenDB: string

outputs:
  krakenResults:
    type: File[]
    outputSource: kraken/results
  krakenTranslateResults:
    type: File[]
    outputSource: krakenTranslate/results

requirements:
  - class: ScatterFeatureRequirement

steps:
  kraken:
    run: kraken.cwl.yml
    in:
      krakenDB: krakenDB
      outputFileName: krakenOutputFileName
      fastaFile: fastaFile
    out: [results]
    scatter: 
      - fastaFile
      - outputFileName
    scatterMethod: dotproduct

  krakenTranslate:
    run: kraken-translate.cwl.yml
    in:
      krakenOutputFile: kraken/results
      krakenDB: krakenDB
      krakenTranslateOutputFileName: krakenTranslateOutputFileName
    out: [results]
    scatter: 
      - krakenOutputFile
      - krakenTranslateOutputFileName
scatterMethod: dotproduct
```

kraken-cwl:
```
cwlVersion: v1.0
class: CommandLineTool
baseCommand: [kraken]

requirements:
 - class: ResourceRequirement
   coresMin: 1
   ramMin: 4096

inputs:
  # krakenDB is a string with a fixed path to the Kraken Databasoe on the
  # shared file system, because Toil doesn't yet support directory-inputs.
  krakenDB:
    type: string
    inputBinding:
      prefix: --db 
      separate: true
      position: 1
  outputFileName:
    type: string
    inputBinding:
      prefix: --output
      separate: true
      position: 2
  fastaFile:
    type: File
    inputBinding:
      position: 3

outputs:
  results:
    type: File
    outputBinding:
glob: $(inputs.outputFileName)
```

kraken-translate-cwl:
```
cwlVersion: v1.0
class: CommandLineTool
baseCommand: [kraken-translate]

requirements:
 - class: ResourceRequirement
   coresMin: 1
   ramMin: 4096

stdout: $(inputs.krakenTranslateOutputFileName)

inputs:
  krakenDB:
    type: string
    inputBinding:
      prefix: --db
      separate: true
      position: 1
  krakenOutputFile:
    type: File
    inputBinding:
      position: 2

outputs:
  results:
type: stdout
```

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-171)
┆friendlyId: TOIL-171

**Comment by ejacox** [comment]
Hello @jsaydo. It looks like the memory allocations are just saying that a job is waiting for other jobs to finish in order to have sufficient memory to run. It looks like there were now errors and everything should have finished correctly.

Also, the log messages are from the single node batch system. Did you use --batchSystem to specify a batch system?

**Comment by jsaydo** [comment]
Hi @ejacox, if I recall it correctly I was using --batchSystem gridengine for the run above. 
I've modified the workflow to a more linear approach and use scattering only in the first step. I had those errors again but somehow now my linear approach works, but I don't understand why. We had some issues with our OpenStack and I had to set up the SGE several times on different instances, so something could have gone wrong there before.
The more parallel approch which is described above still does not work, but I'll look into that later.
